pika/adapters/tornado_connection.py

```diff
@@ -72,7 +73,10 @@ def _adapter_disconnect(self):
         """Disconnect from the RabbitMQ broker"""
         if self.socket:
             self.ioloop.remove_handler(self.socket.fileno())
-        super(TornadoConnection, self)._adapter_disconnect()
+        try:
+            super(TornadoConnection, self)._adapter_disconnect()
+        except exceptions.AMQPConnectionError as e:
```

Comment:
I wonder if the following might do the trick for you (in addition to suppressing AMQPConnectionError in tornado_connection.py):

1. Add try/finally in `BaseConnection._handle_disconnect()`:
 ```
    def _handle_disconnect(self):
        """Called internally when the socket is disconnected already
        """
        try:
            self._adapter_disconnect()
        finally:
            self._on_connection_closed(None, True)
```
2. Pass `on_close_callback` callback function to the TornadoConnection constructor.

I think this will assure that your `on_close_callback` will be called if AMQP handshake fails. The `on_open_error_callback` callback should be called if connection setup (TCP/IP connection or SSL handshake) fails prior to AMQP handshake.